### PR TITLE
Release 10.0.0-alpha.2 -- cleanup, twig prep, typehints

### PR DESCRIPTION
**Pull request by briskt**
### Added
- Added an announcement file for local development
- Added a test feature that logs the html and saves a screenshot on failure

### Changed (breaking)
- Always use the `material` theme
- Added PHP typehints to module code

### Changed (non-breaking)
- Moved PHP code out of templates in preparation for conversion to Twig templating
- Changed the name of the Actions test container to `test`
- Include dev dependencies (e.g. behat) in the test image
- Made the `ADMIN_EMAIL` variable no longer required
- Updated files "lifted" from the SimpleSAMLphp library

### Removed
- Removed `default` theme files
- Removed xdebug from the local development image
- Removed old `prompt-for-mfa-u2f.php` template file
- Removed duplicate copy of simplewebauthn browser.js file

### Fixed
- Fixed a "not found" error on the simplewebauthn browser.js file, which prevented use of a Webauthn 2sv option at login
- Fixed failure in hub mode due to a missing dependency (AWS SDK)
- Fixed failure to ~run~ complete migrations. Reverted the migration namespace back to the original (\Sil\SilAuth\migrations).
- Better error handling for manager "rescue" code request failure

### CODE REVIEW

**Comment by briskt** [comment]
Most changes have been reviewed on a feature branch. Recent changes, such as bugs found while testing on the appsdev IdP and staging hub were committed to the develop branch. Reference [this diff](https://github.com/silinternational/ssp-base/compare/5d884b..b33f09) to see those changes isolated.

File: .github/workflows/test-and-publish.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ jobs:
       - name: Checkout code
         uses: actions/checkout@v4
       - name: test
-        run: docker-compose -f actions-services.yml run --rm app ./run-tests.sh
+        run: docker-compose -f actions-services.yml run --rm test ./run-tests.sh
```
[the remark this answers is not in the document]
Comment (briskt) [reply .github/workflows/test-and-publish.yml:14]: Thanks for the reminder, but I've been aware of that for several years now. It's somewhere down the list of things to address. By the way, by adding this comment you took the php-devs group off of the review list. I had to go add it back.

File: development/idp-local/config/config.php [hunk 2]
```diff
@@ -22,7 +22,7 @@
 
 try {
     // Required to be defined in environment variables
-    $ADMIN_EMAIL = Env::requireEnv('ADMIN_EMAIL');
+    $ADMIN_EMAIL = Env::get('ADMIN_EMAIL', 'na@example.org');
```
Comment (mtompset) [inline development/idp-local/config/config.php:25]: Why this change?

File: development/idp-local/config/config.php [hunk 3]
```diff
@@ -22,7 +22,7 @@
 
 try {
     // Required to be defined in environment variables
```
Comment (mtompset) [inline development/idp-local/config/config.php:24]: Comment is no longer accurate

File: development/idp-local/config/config.php [hunk 4]
```diff
@@ -964,7 +963,7 @@
     /*
      * Which theme directory should be used?
      */
-    'theme.use' => $THEME_USE,
+    'theme.use' => 'material:material',
```
Comment (mtompset) [inline development/idp-local/config/config.php:966]: Why move from variable to hard-coded?

File: development/idp2-local/config/config.php [hunk 5]
```diff
@@ -22,7 +22,7 @@
 
 try {
     // Required to be defined in environment variables
```
Comment (mtompset) [inline development/idp2-local/config/config.php:24]: Comment is inaccurate.

File: development/idp2-local/config/config.php [hunk 6]
```diff
@@ -964,7 +963,7 @@
     /*
      * Which theme directory should be used?
      */
-    'theme.use' => $THEME_USE,
+    'theme.use' => 'material:material',
```
Comment (mtompset) [inline development/idp2-local/config/config.php:966]: Having the variable makes more sense to me. Why this change again?

File: docker-compose.yml [hunk 7]
```diff
@@ -19,7 +19,10 @@ services:
       PMA_PASSWORD: silauth
 
   test:
-    build: .
+    build:
+      context: .
+      args:
```
Comment (mtompset) [inline docker-compose.yml:24]: This was cool to learn about.

File: development/idp-local/config/config.php (hunk 2 above)
Comment (forevermatt) [inline development/idp-local/config/config.php:25]: As @mtompset noticed, since this one isn't required now, it doesn't match the comment explaining this section of the code. Could you move this `ADMIN_EMAIL` line down to somewhere around line 47, to be with the other optional env. vars?

Similarly in the other places an `Env::requireEnv()` was changed to an `Env::get()` with a fallback/default value.

File: features/bootstrap/FeatureContext.php [hunk 8]
```diff
@@ -42,11 +44,27 @@ public function __construct()
     /** @AfterStep */
     public function afterStep(AfterStepScope $scope)
     {
-        if (! $scope->getTestResult()->getResultCode() === StepResult::FAILED) {
+        if ($scope->getTestResult()->getResultCode() === StepResult::FAILED) {
             $this->showPageDetails();
+            $this->takeScreenshot();
```
Comment (forevermatt) [inline features/bootstrap/FeatureContext.php:49]: Nice! 😁

File: modules/mfa/www/prompt-for-mfa.php [hunk 9]
```diff
@@ -124,7 +125,7 @@
 $t->data['stateId'] = $stateId;
 $t->data['supportsWebAuthn'] = LoginBrowser::supportsWebAuthn($userAgent);
 $browserJsHash = md5_file(__DIR__ . '/simplewebauthn/browser.js');
-$t->data['browserJsPath'] = 'simplewebauthn/browser.js?v=' . $browserJsHash;
+$t->data['browserJsPath'] = '/module.php/mfa/simplewebauthn/browser.js?v=' . $browserJsHash;
```
Comment (forevermatt) [inline modules/mfa/www/prompt-for-mfa.php:128]: Was the path still correct in the line calculating the `$browserJsHash`?

**Review by forevermatt — APPROVED** [review]
Approved, with very minor (non-blocking) comments or questions. Thanks! 👍

File: modules/mfa/www/prompt-for-mfa.php (hunk 9 above)
Comment (briskt) [reply modules/mfa/www/prompt-for-mfa.php:128]: Yeah, it's a bit confusing because the path for the controller is different than the path that the browser sees.

File: development/idp-local/config/config.php (hunk 4 above)
Comment (briskt) [reply development/idp-local/config/config.php:965]: Because we never actually use the default template, which is the only other option for our custom modules. It only led to confusion because some tests used "default" and others used "material". Plus, it allowed me to delete a bunch of files related to the default template.

File: development/idp2-local/config/config.php (hunk 6 above)
Comment (briskt) [reply development/idp2-local/config/config.php:965]: Because we never actually use the default template, which is the only other option for our custom modules. It only led to confusion because some tests used "default" and others used "material". Plus, it allowed me to delete a bunch of files related to the default template.
